src/services/db/GitFileSystemService.ts

```diff
@@ -623,17 +631,176 @@ export default class GitFileSystemService {
-      typeof c.date === "string" &&
-      typeof c.message === "string" &&
-      typeof c.hash === "string"
-    )
+  // Rename a single file or directory
+  renameSinglePath(
+    repoName: string,
+    oldPath: string,
+    newPath: string,
+    userId: string,
+    message?: string
+  ): ResultAsync<string, GitFileSystemError | ConflictError> {
+    let oldStateSha = ""
+
+    return this.getLatestCommitOfBranch(repoName, BRANCH_REF)
+      .andThen((latestCommit) => {
+        // It is guaranteed that the latest commit contains the SHA hash
+        oldStateSha = latestCommit.sha as string
+        return okAsync(true)
+      })
+      .andThen(() => this.getFilePathStats(repoName, oldPath))
+      .andThen(() =>
+        // We expect to see an error here, since the new path should not exist
+        this.getFilePathStats(repoName, newPath)
+          .andThen(() =>
+            errAsync(new ConflictError("File path already exists"))
+          )
+          .map(() => true)
+          .orElse((error) => {
+            if (error instanceof NotFoundError) {
+              return okAsync(true)
+            }
+
+            return errAsync(error)
+          })
+      )
+      .andThen(() =>
+        ResultAsync.fromPromise(
+          fs.promises.rename(
+            `${EFS_VOL_PATH}/${repoName}/${oldPath}`,
+            `${EFS_VOL_PATH}/${repoName}/${newPath}`
+          ),
+          (error) => {
+            logger.error(`Error when moving ${oldPath} to ${newPath}: ${error}`)
+
+            if (error instanceof Error) {
+              return new GitFileSystemNeedsRollbackError(
+                `Unable to rename ${oldPath} to ${newPath}`
+              )
+            }
+
+            return new GitFileSystemNeedsRollbackError(
+              "An unknown error occurred"
+            )
+          }
+        )
+      )
+      .andThen(() =>
+        this.commit(
+          repoName,
+          [oldPath, newPath],
+          userId,
+          message || `Renamed ${oldPath} to ${newPath}`
+        )
+      )
+      .orElse((error) => {
+        if (error instanceof GitFileSystemNeedsRollbackError) {
+          return this.rollback(repoName, oldStateSha).andThen(() =>
+            errAsync(new GitFileSystemError(error.message))
+          )
+        }
+
+        return errAsync(error)
+      })
+  }
+
+  // Move multiple files from oldPath to newPath without renaming them
+  moveFiles(
+    repoName: string,
+    oldPath: string,
+    newPath: string,
+    userId: string,
+    targetFiles: string[],
+    message?: string
+  ): ResultAsync<string, GitFileSystemError | ConflictError> {
+    let oldStateSha = ""
+
+    return this.getLatestCommitOfBranch(repoName, BRANCH_REF)
+      .andThen((latestCommit) => {
+        // It is guaranteed that the latest commit contains the SHA hash
+        oldStateSha = latestCommit.sha as string
+        return okAsync(true)
+      })
+      .andThen(() => this.getFilePathStats(repoName, oldPath))
```

Comment:
same as above

Reply:
Ah good point, added in 2f88ef35fc73d0b68d48a66b779f0c4327c70f80!